controllers/infrastructure/byoadmission_controller.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	certv1 "k8s.io/api/certificates/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// ByoAdmissionReconciler reconciles a ByoAdmission object
+type ByoAdmissionReconciler struct {
+	ClientSet clientset.Interface
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=byohosts,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=byohosts/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io,resources=byohosts/finalizers,verbs=update
+//+kubebuilder:rbac:groups=certificates.k8s.io,resources=certificatesigningrequests,verbs=create;get;list;watch
+
+// Reconcile continuosuly checks for CSRs and approves them
+func (r *ByoAdmissionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	logger := log.FromContext(ctx)
+	logger.Info("Reconcile request received", "object", req.NamespacedName)
+
+	// Fetch the CSR from the api-server
+	csr, err := r.ClientSet.CertificatesV1().CertificateSigningRequests().Get(ctx, req.NamespacedName.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Error(err, "CertificateSigningRequest not found, won't reconcile")
+			return reconcile.Result{}, errors.New("CertificateSigningRequest not found")
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Check if the CSR is already approved or denied
+	if checkCSRCondition(csr.Status.Conditions, certv1.CertificateApproved) {
+		return ctrl.Result{}, fmt.Errorf("CertificateSigningRequest %s is already approved", csr.Name)
```

Comment:
Do we have to return error? Can we just log this message and return nil?

Reply:
Currently, we are creating a fake Denied CSR request, checking if during reconciliation we get this error. If it won't return an error, how should we test this?

Reply:
But this isn't really an error, right? Your certificate is already approved, so we should just log it and return nil (so that the request doesn't get re-queued).

For the test, you can assert on returning nil and maybe also on the logged message?

Reply:
Yeah asserted the error to be nil. 
Not sure about asserting on logs with ginkgo. Might require a new buffer for logs for assertion. Tests seem to be good without that.